Simplify indexing when doing ModularIndexing + index propagation.

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119884
* #119857
* #119864
* __->__ #119863

We now avoid creating an unnecessary ternary operator in some reasonably
common case.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames

lezcano committed:
Simplify indexing when doing ModularIndexing + index propagation.

We now avoid creating an unnecessary ternary operator in some reasonably
common case.

[ghstack-poisoned]

lezcano committed:
Update on "Simplify indexing when doing ModularIndexing + index propagation."


We now avoid creating an unnecessary ternary operator in some reasonably
common case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/index_propagation.py:
```diff
@@ -258,5 +258,13 @@ def indirect_indexing(
             # If we are turning a indirect indexing into direct, we need to wrap it.
             index = index.value.expr
+            # If SymPy can prove that idx >= 0, the Where below will evaluate to zero
+            # Alas, SymPy can't prove that index >= when we do ModularIndexing...
```
I think you should put this in `ModularIndexing._eval_is_positive` so sympy will be able to infer this everywhere. Also this should probably be `nonnegative` rather than `positive` since `ModularIndexing` can give zero.

lezcano replied:
ahhhh, that's the function I was looking for. I was just about to ping Aaron to ask him whether this could be generalised.

lezcano committed:
Update on "Simplify indexing when doing ModularIndexing + index propagation."


We now avoid creating an unnecessary ternary operator in some reasonably
common case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

lezcano committed:
Update on "Simplify indexing when doing ModularIndexing + index propagation."


We now avoid creating an unnecessary ternary operator in some reasonably
common case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]
